File: experiments/engine-kt/tools/src/main/kotlin/com/garive/eng/kt/tools/BrowserTools.kt
package com.garive.eng.kt.tools

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** Exact T2 Browser tool revision. */
public const val T2_BROWSER_TOOL_REVISION: String = "1"
/** Pure T2 Browser binding implementation revision. */
public const val T2_BROWSER_RESOLVER_REVISION: String = "garive.t2.browser.access.v1"
/** Observe one admitted browser page. */
public const val T2_BROWSER_OBSERVE: String = "garive.browser.observe"
/** Navigate one admitted browser page. */
public const val T2_BROWSER_NAVIGATE: String = "garive.browser.navigate"
/** Perform one bounded semantic browser action. */
public const val T2_BROWSER_ACT: String = "garive.browser.act"

private const val BROWSER_MAX_RESULT_BYTES: Long = 2_097_152

/** Validated exact browser session/page identity admitted by a catalogue. */
public class BrowserPageScope private constructor(
    private val sessionId: String,
    private val pageId: String,
) {
    internal val resourceKey: String = "browser:$sessionId:$pageId"

    public companion object {
        /** Constructs one scope from portable non-empty ASCII identifiers. */
        public fun create(sessionId: String, pageId: String): ToolContractResult<BrowserPageScope> =
            if (browserToken(sessionId) && browserToken(pageId)) {
                ToolContractResult.Success(BrowserPageScope(sessionId, pageId))
            } else {
                browserFailure()
            }
    }
}

/** Frozen three-tool Browser catalogue for one exact capability snapshot. */
public class BuiltinT2BrowserCatalogue private constructor(
    definitions: List<ToolDefinition>,
    private val catalogue: ToolCatalog,
) {
    /** Exact definitions frozen into the Agent snapshot. */
    public val definitions: List<ToolDefinition> = definitions.toList()

    /** Prepares one Browser intent through exact page and origin bindings. */
    public fun prepare(intent: ToolIntent): ToolContractResult<PreparedToolCall> =
        catalogue.prepareV3(intent, BrowserResolver(intent.toolName))

    public companion object {
        /** Freezes exact admitted pages, canonical origins and policy revision. */
        public fun create(
            policyRevision: String,
            pages: List<BrowserPageScope>,
            origins: List<String>,
        ): ToolContractResult<BuiltinT2BrowserCatalogue> = browserContract {
            val definitions = listOf(
                browserObserveDefinition(policyRevision, pages),
                browserNavigateDefinition(policyRevision, pages, origins),
                browserActDefinition(policyRevision, pages, origins),
            ).sortedBy(ToolDefinition::name)
            BuiltinT2BrowserCatalogue(definitions, ToolCatalog.create(definitions).browserRequired())
        }
    }
}

private class BrowserResolver(private val toolName: String) : ToolAccessResolver {
    override val revision: String = T2_BROWSER_RESOLVER_REVISION

    override fun resolve(arguments: JsonElement): ToolContractResult<InvocationAccessSet> = browserContract {
        val page = BrowserPageScope.create(arguments.browserText("session_id"), arguments.browserText("page_id"))
            .browserRequired()
        val mode = if (toolName == T2_BROWSER_OBSERVE) AccessMode.READ else AccessMode.WRITE
        val accesses = mutableListOf(
            ResourceAccess.create(AccessNamespace.RUNTIME, page.resourceKey, mode).browserRequired(),
        )
        when (toolName) {
            T2_BROWSER_OBSERVE -> Unit
            T2_BROWSER_NAVIGATE -> {
                val origin = arguments.browserText("destination_origin")
                if (browserUrlOrigin(arguments.browserText("destination_url")) != origin) browserAbort()
                accesses += ResourceAccess.create(AccessNamespace.NETWORK, origin, AccessMode.WRITE).browserRequired()
            }
            T2_BROWSER_ACT -> {
                validateBrowserAction(arguments)
                arguments.browserStrings("allowed_navigation_origins").forEach { origin ->
                    accesses += ResourceAccess.create(AccessNamespace.NETWORK, origin, AccessMode.WRITE)
                        .browserRequired()
                }
            }
            else -> browserAbort()
        }
        InvocationAccessSet.create(accesses).browserRequired()
    }
}

private fun browserObserveDefinition(policy: String, pages: List<BrowserPageScope>): ToolDefinition =
    browserDefinition(
        T2_BROWSER_OBSERVE,
        "Observe one bounded browser semantic tree.",
        browserSchema("""{"type":"object","properties":{"session_id":{"type":"string","minLength":1,"maxLength":128},"page_id":{"type":"string","minLength":1,"maxLength":128},"expected_previous_snapshot_id":{"type":"string","minLength":1,"maxLength":128},"max_nodes":{"type":"integer","minimum":1,"maximum":10000},"max_text_bytes":{"type":"integer","minimum":1,"maximum":1048576}},"required":["session_id","page_id","max_nodes","max_text_bytes"],"additionalProperties":false}"""),
        listOf(ExecutionCapability.BROWSER_OBSERVE),
        ReplayClass.READ_ONLY,
        pages,
        emptyList(),
        AccessMode.READ,
        policy,
    )

private fun browserNavigateDefinition(
    policy: String,
    pages: List<BrowserPageScope>,
    origins: List<String>,
): ToolDefinition = browserDefinition(
    T2_BROWSER_NAVIGATE,
    "Navigate one browser page to an exact admitted origin.",
    browserSchema("""{"type":"object","properties":{"session_id":{"type":"string","minLength":1,"maxLength":128},"page_id":{"type":"string","minLength":1,"maxLength":128},"expected_snapshot_id":{"type":"string","minLength":1,"maxLength":128},"destination_url":{"type":"string","minLength":1,"maxLength":8192},"destination_origin":{"type":"string","minLength":10,"maxLength":512},"wait_until":{"type":"string","enum":["dom_content_loaded","load","network_idle"]},"timeout_ms":{"type":"integer","minimum":1,"maximum":120000},"max_nodes":{"type":"integer","minimum":1,"maximum":10000},"max_text_bytes":{"type":"integer","minimum":1,"maximum":1048576}},"required":["session_id","page_id","expected_snapshot_id","destination_url","destination_origin","wait_until","timeout_ms","max_nodes","max_text_bytes"],"additionalProperties":false}"""),
    listOf(ExecutionCapability.BROWSER_ACT, ExecutionCapability.NETWORK),
    ReplayClass.NEVER_REPLAY,
    pages,
    origins,
    AccessMode.WRITE,
    policy,
)

private fun browserActDefinition(
    policy: String,
    pages: List<BrowserPageScope>,
    origins: List<String>,
): ToolDefinition = browserDefinition(
    T2_BROWSER_ACT,
    "Perform one snapshot-bound semantic browser action.",
    browserSchema("""{"type":"object","properties":{"session_id":{"type":"string","minLength":1,"maxLength":128},"page_id":{"type":"string","minLength":1,"maxLength":128},"expected_snapshot_id":{"type":"string","minLength":1,"maxLength":128},"action":{"type":"string","enum":["click","type_text","clear","select_option","press_key","scroll","go_back","go_forward","reload"]},"node_ref":{"type":"string","minLength":1,"maxLength":128},"text":{"type":"string","maxLength":32768},"option":{"type":"string","minLength":1,"maxLength":4096},"key":{"type":"string","enum":["enter","tab","escape","backspace","delete","arrow_up","arrow_down","arrow_left","arrow_right","home","end","page_up","page_down","space"]},"delta_x":{"type":"integer","minimum":-100000,"maximum":100000},"delta_y":{"type":"integer","minimum":-100000,"maximum":100000},"allowed_navigation_origins":{"type":"array","maxItems":16,"items":{"type":"string","minLength":10,"maxLength":512}}},"required":["session_id","page_id","expected_snapshot_id","action","allowed_navigation_origins"],"additionalProperties":false}"""),
    listOf(ExecutionCapability.BROWSER_ACT, ExecutionCapability.NETWORK),
    ReplayClass.NEVER_REPLAY,
    pages,
    origins,
    AccessMode.WRITE,
    policy,
)

@Suppress("LongParameterList")
private fun browserDefinition(
    name: String,
    description: String,
    inputSchema: JsonElement,
    capabilities: List<ExecutionCapability>,
    replay: ReplayClass,
    pages: List<BrowserPageScope>,
    origins: List<String>,
    pageMode: AccessMode,
    policy: String,
): ToolDefinition {
    val observe = name == T2_BROWSER_OBSERVE
    val requirements = ExecutionRequirements.create(
        capabilities,
        if (observe) 30_000 else 120_000,
        BROWSER_MAX_RESULT_BYTES,
    ).browserRequired()
    val controls = mutableListOf(
        SandboxControl.BROWSER_SESSION_SCOPE,
        SandboxControl.SNAPSHOT_BINDING,
        SandboxControl.RESOURCE_LIMITS,
    )
    if (ExecutionCapability.NETWORK in capabilities) {
        controls += listOf(SandboxControl.NETWORK_ORIGIN_SCOPE, SandboxControl.REDIRECT_REVALIDATION)
    }
    return ToolDefinition.createV3(
        name,
        T2_BROWSER_TOOL_REVISION,
        description,
        inputSchema,
        requirements,
        replay,
        ToolAccessPolicyV1.create(
            policy,
            emptyList(),
            emptyList(),
            origins.map { AccessPolicyEntry.create(it, listOf(AccessMode.WRITE)).browserRequired() },
            pages.map { AccessPolicyEntry.create(it.resourceKey, listOf(pageMode)).browserRequired() },
            if (observe) 1 else 17,
            BROWSER_MAX_RESULT_BYTES,
        ).browserRequired(),
        T2_BROWSER_RESOLVER_REVISION,
        SandboxRequirementsV1.create(capabilities, controls, null, 64).browserRequired(),
    ).browserRequired()
}

private fun validateBrowserAction(arguments: JsonElement) {
    val value = arguments.jsonObject
    fun present(name: String): Boolean = name in value
    val valid = when (arguments.browserText("action")) {
        "click", "clear" -> present("node_ref") && absent(value, "text", "option", "key", "delta_x", "delta_y")
        "type_text" -> present("node_ref") && present("text") && absent(value, "option", "key", "delta_x", "delta_y")
        "select_option" -> present("node_ref") && present("option") && absent(value, "text", "key", "delta_x", "delta_y")
        "press_key" -> present("key") && absent(value, "node_ref", "text", "option", "delta_x", "delta_y")
        "scroll" -> present("delta_x") && present("delta_y") &&
            (value.getValue("delta_x").jsonPrimitive.int != 0 || value.getValue("delta_y").jsonPrimitive.int != 0) &&
            absent(value, "node_ref", "text", "option", "key")
        "go_back", "go_forward", "reload" -> absent(value, "node_ref", "text", "option", "key", "delta_x", "delta_y")
        else -> false
    }
    if (!valid) browserAbort()
}

private fun absent(value: JsonObject, vararg names: String): Boolean = names.none(value::containsKey)
private fun browserUrlOrigin(url: String): String? {
    val split = url.indexOf("://")
    if (split < 0 || url.substring(0, split) !in setOf("http", "https")) return null
    val end = url.indexOfAny(charArrayOf('/', '?', '#'), split + 3).let { if (it < 0) url.length else it }
    return url.substring(0, end)
}
private fun browserToken(value: String): Boolean = value.isNotEmpty() && value.length <= 128 &&
    value.all { (it.isLetterOrDigit() && it.code < 128) || it in "-_." }
private fun JsonElement.browserText(name: String): String =
    (this as? JsonObject)?.get(name)?.jsonPrimitive?.content ?: browserAbort()
private fun JsonElement.browserStrings(name: String): List<String> =
    (this as? JsonObject)?.get(name)?.let { it as? JsonArray }?.map { it.jsonPrimitive.content } ?: browserAbort()
private fun browserSchema(value: String): JsonElement = Json.parseToJsonElement(value)
private class BrowserContractFailure(val error: PreparationError) : RuntimeException()
private fun browserAbort(): Nothing = throw BrowserContractFailure(PreparationError(PreparationErrorCode.EFFECT_ACCESS_INVALID))
private fun browserFailure(): ToolContractResult.Failure =
    ToolContractResult.Failure(PreparationError(PreparationErrorCode.EFFECT_ACCESS_INVALID))
private fun <T> ToolContractResult<T>.browserRequired(): T = when (this) {
    is ToolContractResult.Success -> value
    is ToolContractResult.Failure -> throw BrowserContractFailure(error)
}
private inline fun <T> browserContract(block: () -> T): ToolContractResult<T> = try {
    ToolContractResult.Success(block())
} catch (failure: BrowserContractFailure) {
    ToolContractResult.Failure(failure.error)
}

File: engine/tools/src/browser_tools.rs
//! Exact T2 browser definitions and pure session/origin bindings.

use serde_json::{json, Value};

use crate::{
    AccessMode, AccessNamespace, AccessPolicyEntry, ExecutionCapability, ExecutionRequirements,
    InvocationAccessSet, PreparationError, PreparationErrorCode, PreparedToolCall, ReplayClass,
    ResourceAccess, SandboxControl, SandboxRequirementsV1, ToolAccessPolicyV1, ToolAccessResolver,
    ToolCatalog, ToolDefinition, ToolIntent,
};

/// Exact T2 Browser tool revision.
pub const T2_BROWSER_TOOL_REVISION: &str = "1";
/// Pure T2 Browser binding implementation revision.
pub const T2_BROWSER_RESOLVER_REVISION: &str = "garive.t2.browser.access.v1";
/// Observe one admitted browser page.
pub const T2_BROWSER_OBSERVE: &str = "garive.browser.observe";
/// Navigate one admitted browser page.
pub const T2_BROWSER_NAVIGATE: &str = "garive.browser.navigate";
/// Perform one bounded semantic browser action.
pub const T2_BROWSER_ACT: &str = "garive.browser.act";

const MAX_RESULT_BYTES: u64 = 2_097_152;

/// Validated exact browser session/page identity admitted by a catalogue.
#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
pub struct BrowserPageScope {
    session_id: String,
    page_id: String,
}

impl BrowserPageScope {
    /// Constructs one scope from portable non-empty ASCII identifiers.
    pub fn new(
        session_id: impl Into<String>,
        page_id: impl Into<String>,
    ) -> Result<Self, PreparationError> {
        let value = Self {
            session_id: session_id.into(),
            page_id: page_id.into(),
        };
        if !token(&value.session_id) || !token(&value.page_id) {
            return Err(access_error());
        }
        Ok(value)
    }

    fn resource_key(&self) -> String {
        format!("browser:{}:{}", self.session_id, self.page_id)
    }
}

/// Frozen three-tool Browser catalogue for one exact capability snapshot.
#[derive(Clone, Debug)]
pub struct BuiltinT2BrowserCatalogue {
    definitions: Vec<ToolDefinition>,
    catalog: ToolCatalog,
}

impl BuiltinT2BrowserCatalogue {
    /// Freezes exact admitted pages, canonical origins and policy revision.
    pub fn new(
        policy_revision: impl Into<String>,
        pages: impl IntoIterator<Item = BrowserPageScope>,
        origins: impl IntoIterator<Item = impl Into<String>>,
    ) -> Result<Self, PreparationError> {
        let policy_revision = policy_revision.into();
        let pages = pages.into_iter().collect::<Vec<_>>();
        let origins = origins.into_iter().map(Into::into).collect::<Vec<_>>();
        let mut definitions = vec![
            observe_definition(&policy_revision, &pages)?,
            navigate_definition(&policy_revision, &pages, &origins)?,
            act_definition(&policy_revision, &pages, &origins)?,
        ];
        definitions.sort_by(|left, right| left.name().cmp(right.name()));
        let catalog = ToolCatalog::new(definitions.clone())?;
        Ok(Self {
            definitions,
            catalog,
        })
    }

    /// Returns the exact definitions frozen into the Agent snapshot.
    pub fn definitions(&self) -> &[ToolDefinition] {
        &self.definitions
    }

    /// Prepares one Browser intent through exact page and origin bindings.
    pub fn prepare(&self, intent: &ToolIntent) -> Result<PreparedToolCall, PreparationError> {
        self.catalog.prepare_v3(
            intent,
            &BrowserResolver {
                tool_name: intent.tool_name(),
            },
        )
    }
}

struct BrowserResolver<'a> {
    tool_name: &'a str,
}

impl ToolAccessResolver for BrowserResolver<'_> {
    fn revision(&self) -> &str {
        T2_BROWSER_RESOLVER_REVISION
    }

    fn resolve(&self, arguments: &Value) -> Result<InvocationAccessSet, PreparationError> {
        let page =
            BrowserPageScope::new(text(arguments, "session_id")?, text(arguments, "page_id")?)?;
        let mode = if self.tool_name == T2_BROWSER_OBSERVE {
            AccessMode::Read
        } else {
            AccessMode::Write
        };
        let mut accesses = vec![ResourceAccess::new(
            AccessNamespace::Runtime,
            page.resource_key(),
            mode,
        )?];
        match self.tool_name {
            T2_BROWSER_OBSERVE => {}
            T2_BROWSER_NAVIGATE => {
                let origin = text(arguments, "destination_origin")?;
                if url_origin(text(arguments, "destination_url")?) != Some(origin) {
                    return Err(access_error());
                }
                accesses.push(ResourceAccess::new(
                    AccessNamespace::Network,
                    origin,
                    AccessMode::Write,
                )?);
            }
            T2_BROWSER_ACT => {
                validate_action(arguments)?;
                for origin in string_array(arguments, "allowed_navigation_origins")? {
                    accesses.push(ResourceAccess::new(
                        AccessNamespace::Network,
                        origin,
                        AccessMode::Write,
                    )?);
                }
            }
            _ => return Err(access_error()),
        }
        InvocationAccessSet::new(accesses)
    }
}

fn observe_definition(
    policy: &str,
    pages: &[BrowserPageScope],
) -> Result<ToolDefinition, PreparationError> {
    definition(
        T2_BROWSER_OBSERVE,
        "Observe one bounded browser semantic tree.",
        json!({"type":"object","properties":{"session_id":{"type":"string","minLength":1,"maxLength":128},"page_id":{"type":"string","minLength":1,"maxLength":128},"expected_previous_snapshot_id":{"type":"string","minLength":1,"maxLength":128},"max_nodes":{"type":"integer","minimum":1,"maximum":10000},"max_text_bytes":{"type":"integer","minimum":1,"maximum":1048576}},"required":["session_id","page_id","max_nodes","max_text_bytes"],"additionalProperties":false}),
        [ExecutionCapability::BrowserObserve],
        ReplayClass::ReadOnly,
        pages,
        &[],
        AccessMode::Read,
        policy,
    )
}

fn navigate_definition(
    policy: &str,
    pages: &[BrowserPageScope],
    origins: &[String],
) -> Result<ToolDefinition, PreparationError> {
    definition(
        T2_BROWSER_NAVIGATE,
        "Navigate one browser page to an exact admitted origin.",
        json!({"type":"object","properties":{"session_id":{"type":"string","minLength":1,"maxLength":128},"page_id":{"type":"string","minLength":1,"maxLength":128},"expected_snapshot_id":{"type":"string","minLength":1,"maxLength":128},"target_revision":{"type":"string","minLength":1,"maxLength":128},"destination_url":{"type":"string","minLength":1,"maxLength":8192},"destination_origin":{"type":"string","minLength":10,"maxLength":512},"wait_until":{"type":"string","enum":["dom_content_loaded","load","network_idle"]},"timeout_ms":{"type":"integer","minimum":1,"maximum":120000},"max_nodes":{"type":"integer","minimum":1,"maximum":10000},"max_text_bytes":{"type":"integer","minimum":1,"maximum":1048576}},"required":["session_id","page_id","expected_snapshot_id","target_revision","destination_url","destination_origin","wait_until","timeout_ms","max_nodes","max_text_bytes"],"additionalProperties":false}),
        [
            ExecutionCapability::BrowserAct,
            ExecutionCapability::Network,
        ],
        ReplayClass::NeverReplay,
        pages,
        origins,
        AccessMode::Write,
        policy,
    )
}

fn act_definition(
    policy: &str,
    pages: &[BrowserPageScope],
    origins: &[String],
) -> Result<ToolDefinition, PreparationError> {
    definition(
        T2_BROWSER_ACT,
        "Perform one snapshot-bound semantic browser action.",
        json!({"type":"object","properties":{"session_id":{"type":"string","minLength":1,"maxLength":128},"page_id":{"type":"string","minLength":1,"maxLength":128},"expected_snapshot_id":{"type":"string","minLength":1,"maxLength":128},"target_revision":{"type":"string","minLength":1,"maxLength":128},"action":{"type":"string","enum":["click","type_text","clear","select_option","press_key","scroll","go_back","go_forward","reload"]},"node_ref":{"type":"string","minLength":1,"maxLength":128},"text":{"type":"string","maxLength":32768},"option":{"type":"string","minLength":1,"maxLength":4096},"key":{"type":"string","enum":["enter","tab","escape","backspace","delete","arrow_up","arrow_down","arrow_left","arrow_right","home","end","page_up","page_down","space"]},"delta_x":{"type":"integer","minimum":-100000,"maximum":100000},"delta_y":{"type":"integer","minimum":-100000,"maximum":100000},"allowed_navigation_origins":{"type":"array","maxItems":16,"items":{"type":"string","minLength":10,"maxLength":512}}},"required":["session_id","page_id","expected_snapshot_id","target_revision","action","allowed_navigation_origins"],"additionalProperties":false}),
        [
            ExecutionCapability::BrowserAct,
            ExecutionCapability::Network,
        ],
        ReplayClass::NeverReplay,
        pages,
        origins,
        AccessMode::Write,
        policy,
    )
}

#[allow(clippy::too_many_arguments)]
fn definition<const C: usize>(
    name: &str,
    description: &str,
    schema: Value,
    capabilities: [ExecutionCapability; C],
    replay: ReplayClass,
    pages: &[BrowserPageScope],
    origins: &[String],
    page_mode: AccessMode,
    policy: &str,
) -> Result<ToolDefinition, PreparationError> {
    let observe = name == T2_BROWSER_OBSERVE;
    let requirements = ExecutionRequirements::new(
        capabilities,
        if observe { 30_000 } else { 120_000 },
        MAX_RESULT_BYTES,
    )?;
    let controls = if capabilities.contains(&ExecutionCapability::Network) {
        vec![
            SandboxControl::NetworkOriginScope,
            SandboxControl::RedirectRevalidation,
            SandboxControl::BrowserSessionScope,
            SandboxControl::SnapshotBinding,
            SandboxControl::ResourceLimits,
        ]
    } else {
        vec![
            SandboxControl::BrowserSessionScope,
            SandboxControl::SnapshotBinding,
            SandboxControl::ResourceLimits,
        ]
    };
    ToolDefinition::new_v3(
        name,
        T2_BROWSER_TOOL_REVISION,
        description,
        schema,
        requirements.clone(),
        replay,
        ToolAccessPolicyV1::new(
            policy,
            [],
            [],
            origins
                .iter()
                .map(|origin| AccessPolicyEntry::new(origin, [AccessMode::Write]))
                .collect::<Result<Vec<_>, _>>()?,
            pages
                .iter()
                .map(|page| AccessPolicyEntry::new(page.resource_key(), [page_mode]))
                .collect::<Result<Vec<_>, _>>()?,
            if observe { 1 } else { 17 },
            MAX_RESULT_BYTES,
        )?,
        T2_BROWSER_RESOLVER_REVISION,
        SandboxRequirementsV1::new(requirements.capabilities(), controls, None, 64)?,
    )
}

fn validate_action(arguments: &Value) -> Result<(), PreparationError> {
    let present = |name: &str| arguments.get(name).is_some();
    let valid = match text(arguments, "action")? {
        "click" | "clear" => {
            present("node_ref")
                && !present("text")
                && !present("option")
                && !present("key")
                && !present("delta_x")
                && !present("delta_y")
        }
        "type_text" => {
            present("node_ref")
                && present("text")
                && !present("option")
                && !present("key")
                && !present("delta_x")
                && !present("delta_y")
        }
        "select_option" => {
            present("node_ref")
                && present("option")
                && !present("text")
                && !present("key")
                && !present("delta_x")
                && !present("delta_y")
        }
        "press_key" => {
            present("key")
                && !present("node_ref")
                && !present("text")
                && !present("option")
                && !present("delta_x")
                && !present("delta_y")
        }
        "scroll" => {
            present("delta_x")
                && present("delta_y")
                && (integer(arguments, "delta_x")? != 0 || integer(arguments, "delta_y")? != 0)
                && !present("node_ref")
                && !present("text")
                && !present("option")
                && !present("key")
        }
        "go_back" | "go_forward" | "reload" => {
            !["node_ref", "text", "option", "key", "delta_x", "delta_y"]
                .iter()
                .any(|name| present(name))
        }
        _ => false,
    };
    if !valid {
        return Err(access_error());
    }
    Ok(())
}

fn string_array<'a>(value: &'a Value, field: &str) -> Result<Vec<&'a str>, PreparationError> {
    value
        .get(field)
        .and_then(Value::as_array)
        .ok_or_else(access_error)?
        .iter()
        .map(|value| value.as_str().ok_or_else(access_error))
        .collect()
}

fn integer(value: &Value, field: &str) -> Result<i64, PreparationError> {
    value
        .get(field)
        .and_then(Value::as_i64)
        .ok_or_else(access_error)
}

fn url_origin(url: &str) -> Option<&str> {
    let scheme_end = url.find("://")?;
    if !matches!(&url[..scheme_end], "http" | "https") {
        return None;
    }
    let authority_end = url[scheme_end + 3..]
        .find(['/', '?', '#'])
        .map_or(url.len(), |index| scheme_end + 3 + index);
    Some(&url[..authority_end])
}

fn token(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 128
        && value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.'))
}

fn text<'a>(value: &'a Value, field: &str) -> Result<&'a str, PreparationError> {
    value
        .get(field)
        .and_then(Value::as_str)
        .ok_or_else(access_error)
}

fn access_error() -> PreparationError {
    PreparationError::new(PreparationErrorCode::EffectAccessInvalid)
}
